File: src/sort/java/divideConquer/MergeSort.java
package sort.java.divideConquer;

import java.util.Arrays;

/**
 * Created by shaojiexu on 2/18/18.
 */
public class MergeSort {

    public static void main(String... args) {
        int[] a = {7,8,9,1,2,3,10,23,0};
        mergesort(a);
        System.out.println(Arrays.toString(mergesort(a)));
    }

    private static int[] merge(int[] a, int[] b) {
        int[] c = new int[a.length + b.length];
        int i = 0, j = 0;
        for (int k = 0; k < c.length; k++) {
            if      (i >= a.length) c[k] = b[j++];
            else if (j >= b.length) c[k] = a[i++];
            else if (a[i] <= b[j])  c[k] = a[i++];
            else                    c[k] = b[j++];
        }
        return c;
    }

    public static int[] mergesort(int[] input) {
        int N = input.length;
        if (N <= 1) return input;
        int[] a = new int[N/2];
        int[] b = new int[N - N/2];
        for (int i = 0; i < a.length; i++)
            a[i] = input[i];
        for (int i = 0; i < b.length; i++)
            b[i] = input[i + N/2];
        return merge(mergesort(a), mergesort(b));
    }

}

File: src/sort/scala/divideConquer/MergeSort.scala
package sort.scala.divideConquer

/**
  * Worst - O(n log n)
  * Best - O(n log n)
  * Average - O(n log n)
  *
  */
object MergeSort extends App {


  def merge[T : Ordering](l1: List[T], l2:List[T]) : List[T] = {
    (l1, l2) match {
      case (l1, Nil) => l1
      case (Nil, l2) => l2
      case (l1Head :: l1Tail, l2Head :: l2Tail) =>
        if(implicitly[Ordering[T]].lteq(l1Head, l2Head))  l1Head::merge(l1Tail, l2)
        else l2Head::merge(l1, l2Tail)
    }
  }

  def mergeSort[T : Ordering](l : List[T]) : List[T] = {
    l match {
      case Nil => Nil
      case lhead :: Nil => l
      case lhead :: ltail =>
        val (left, right) = l.splitAt(l.length/2)
        merge(mergeSort(left), mergeSort(right))
    }
  }

  implicit val orderByAge : Ordering[Person] = (p1, p2) => p1.age - p2.age
  implicit val orderByName : Ordering[Person] = (p1, p2) => p1.name.compare(p2.name)

  val employee = List(Person("Jack",21), Person("Angela", 58), Person("Jacobo", 30), Person("Anna", 20), Person("Ian",45))
  println(mergeSort(employee)(orderByAge))
  println(mergeSort(employee)(orderByName))

}
